src/Razor/Microsoft.NET.Sdk.Razor/src/ApplyCssScopes.cs

```diff
@@ -0,0 +1,95 @@
+
+                    if (string.Equals(componentCandidate.ItemSpec, razorComponent, StringComparison.OrdinalIgnoreCase))
+                    {
+                        unmatchedScopedCss.RemoveAt(j);
+                        if (!scopedCssByComponent.TryGetValue(componentCandidate.ItemSpec, out var existing))
+                        {
+                            scopedCssByComponent[componentCandidate.ItemSpec] = new List<ITaskItem>() { scopedCssCandidate };
+                            var item = new TaskItem(componentCandidate);
+                            item.SetMetadata("CssScope", scopedCssCandidate.GetMetadata("CssScope"));
+                            razorComponentsWithScopes.Add(item);
+                        }
+                        else
+                        {
+                            existing.Add(scopedCssCandidate);
+                        }
+                    }
+                    else
+                    {
+                        j++;
+                    }
+                }
+            }
+
+            foreach (var kvp in scopedCssByComponent)
+            {
+                var component = kvp.Key;
+                var scopeFiles = kvp.Value;
+
+                if (scopeFiles.Count > 1)
+                {
+                    Log.LogError($"More than one scoped css files were found for the razor component '{component}'. Each razor component must have at most" +
+                    " a single associated scoped css file." + Environment.NewLine + string.Join(Environment.NewLine, scopeFiles.Select(f => f.ItemSpec)));
+                }
+            }
+
+            // We don't want to allow scoped css files without a matching component. Our convention is very specific in its requirements
+            // so failing to have a matching component very likely means an error.
+            // When the matching component was specified explicitly, failing to find a matching component is an error.
+            // This simplifies a few things like being able to assume that the presence of a .razor.css file or a ScopedCssInput item will result in a bundle being produced,
+            // that the contents of the bundle are independent of the existence of a component and that users will be able to catch errors at compile
+            // time instead of wondering why their component doesn't have a scope applied to it.
+            // In the rare case that a .razor file exists on the user project, has an associated .razor.css file and the user decides to exclude it as a RazorComponent they
+            // can update the Content item for the .razor.css file with Scoped=false and we will not consider it.
+            foreach (var unmatched in unmatchedScopedCss)
+            {
+                Log.LogError($"The scoped css file '{unmatched.ItemSpec}' was defined but no associated razor component was found for it.");
```

Comment:
We can do this in a follow up, but it's useful to include an error code with these errors.

Reply:
What would be an acceptable error code? I tried to search for previous history on this but couldn't find anything.